### PR TITLE
The ghost drop event now fires on player death and disconnect.

### PR DESCRIPTION
**Pull request by Agiel**
Also use m_iGameState instead of round time to determine round end.

### CODE REVIEW

**Comment by Rainyan** [comment]
Does this change take into account that disconnecting client will also destroy any weapon edicts in their possession as they leave the server, the ghost incuded? If not, then you'd have the _OnGhostDrop(...)_ trigger when in actuality the ghost no longer exists in the round.

I would suggest manually dropping the ghost before allowing the _OnClientDisconnect(...)_ to go through:

```diff
--- a/scripting/nt_ghostcap.sp
+++ b/scripting/nt_ghostcap.sp
@@ -80,7 +80,15 @@ public void OnClientDisconnect(int client)
 {
        if (client == ghostCarrier)
        {
-               PushOnGhostDrop(client);
+               int primary_weapon = GetEntPropEnt(client, Prop_Data, "m_hMyWeapons");
+               // Assuming that when (client == ghostCarrier), player's primary weapon must be the ghost.
+               if (primary_weapon != -1)
+               {
+                       float client_pos[3];
+                       GetClientAbsOrigin(client, client_pos);
+                       SDKHooks_DropWeapon(client, primary_weapon, client_pos, NULL_VECTOR);
+                       PushOnGhostDrop(client);
+               }
        }
 }
 ```
 
@Agiel  I would post a PR but my own fork is not currently in sync - perhaps you could take a look?

**Comment by Agiel** [comment]
Good idea. I'm not opposed to putting it here, but it does feel a bit outside the scope for this plugin. Unless you think there are other plugins that will break if we don't make this change?

I'll be busy today, but if everyone agrees I'll try to submit a PR tomorrow.

**Comment by Rainyan** [comment]
TLDR: Probably not as big of a deal as I initially thought. I'd personally go for the option C explained at the bottom of this post, but I'm willing to settle with leaving things as-is.

Any thoughts?

___

> Unless you think there are other plugins that will break if we don't make this change?

Of the top of my head I don't think anyone is relying on ghost actually existing following a _OnGhostDrop_, so I doubt anything would break.

Hypothetical worst case would be someone caching a ghost index, and assuming it's still valid after _OnGhostDrop_ fires, without error checking (although they should probably be using entrefs to catch this kind of thing anyways).

I guess it boils down to whether _OnGhostDrop_ should semantically imply that the ghost has been moved from player to world, or just that the player who previously held it, no longer does. I don't really mind either way, but it just feels a bit unclean if the post-OnGhostDrop tick may or may not have the ghost existing, depending on which code path the event originated from.

It would make sense as a separate plugin, however the problem is that the _OnWeaponDrop_ that would result by it is too late for the disconnecting client to trigger anymore (client disconnect tick goes through, client index becomes invalid, hence _OnWeaponDrop_ doesn't trigger), so nt_ghostcap would have no way of seeing it. You could manually call OnWeaponDrop with a bogus client index, but that would break anyone relying on it always returning valid client indices, so it's probably not an option.

So I guess the main options are:

* **A)** Just fire _OnGhostDrop_ from _OnClientDisconnect_ and ignore whether the ghost exists after it.

Basically the state of this PR. We don't have guarantees on ghost existence on the game tick following _OnGhostDrop_, and maybe that's fine. Does mean error checks are required for confirming the ghost still exists post-OnGhostDrop.

* **B)** Expose the _OnGhostDrop_ as some kind of plugin native (similar to #12 ), so that the separate ghost drop handling plugin could manually trigger _OnGhostDrop_ after handling _OnClientDisconnect_.

This would guarantee that ghost exists in the game tick following _OnGhostDrop_. But it seems a bit overkill, since I doubt any other plugin would ever need such a native outside this one corner case.

* **C)** Handle the ghost rescue inside nt_ghostcap's _OnClientDisconnect_.

Basically the code snippet I posted in the earlier message. It does feel a bit out-of-scope, but I like how it doesn't change our (admittedly nonexplicit) guarantees of ghost existence on tick after _OnGhostDrop_, regardless of where the event originated from. I would probably be most in favour of this, followed by option A.

**Comment by Agiel** [comment]
I agree option C makes the most sense. I verified your code myself and it does work as expected :)

See #14
